addons/l10n_in/models/account_invoice.py

```diff
@@ -71,18 +71,36 @@ def _compute_l10n_in_state_id(self):
 
     @api.depends('l10n_in_state_id')
     def _compute_fiscal_position_id(self):
-        other_country = None
-        in_moves = self.filtered(lambda move: move.country_code == 'IN' and move.is_sale_document(include_receipts=True))
+        foreign_state = False
+        in_moves = self.filtered(lambda move: move.country_code == 'IN' and move.is_invoice(include_receipts=True))
         for move in in_moves:
-            state_id = move.l10n_in_state_id.id
-            country_id = move.l10n_in_state_id.country_id.id
-            if move.l10n_in_state_id.l10n_in_tin == '96':  # Foreign Country
-                other_country = other_country or self.env['res.country'].search([('code', '!=', 'IN')], limit=1)
-                country_id = other_country.id
-                state_id = False
+            state_id = move.l10n_in_state_id
+            is_purchase_document = move.is_purchase_document(include_receipts=True)
+            if (
+                is_purchase_document
+                and state_id.l10n_in_tin != '96'
+                and move.partner_id.country_id.code == 'IN'
+            ):
+                if state_id == move.partner_id.state_id:
+                    # Set Intra State for purchase if place of supply is same as partner state
+                    state_id = move.company_id.state_id
+                elif state_id == move.company_id.state_id:
+                    # To apply inter state tax for purchase if place of supply is same as company state
+                    state_id = self.env['res.country.state'].search([
+                        ('country_id', '=', move.company_id.country_id.id),
+                        ('id', 'not in', [move.company_id.state_id.id, move.partner_id.state_id.id])
+                    ], limit=1)
+            if (
+                state_id.l10n_in_tin == '96'
+                or (is_purchase_document and move.partner_id.country_id.code != 'IN')
+            ):  # Foreign Country
+                foreign_state = foreign_state or self.env['res.country.state'].search([
+                    ('country_id', '!=', move.company_id.country_id.id)
+                ], limit=1)
+                state_id = foreign_state
```

Comment:
No, this should be handled by the fiscal position itself.  We could do country group India and now there is excluded_state_ids on there where we can put the state with tin 96.  (or get rid of state 96)